### DISCUSSION TITLE
Jira Tickets with publishToConfluence

### DISCUSSION THREAD

**Discussion by DaviSchu** [docToolchain/docToolchain#916] [Q&A]
I need some help with how to generate correct Jira Links when I publish a document into Confluence.

The Upload itself is working fine and I am using it already for quite a while. Now I looked into how to create Jira Links. The very first thing I tried, was to simply hand through the original Confluence macro source code within the published document like this:
```html
++++
<ac:structured-macro ac:name="jira" ac:schema-version="1">
    <ac:parameter ac:name="server">myServerName</ac:parameter>
    <ac:parameter ac:name="serverId">myServerId</ac:parameter>
    <ac:parameter ac:name="key">ticketKey</ac:parameter>
</ac:structured-macro>
++++
```
This does only work partly. I do get a Jira Link Macro that looks like every other one I can create myself in Confluence directly. I can edit and work with it normally. It also redirects to the correct Jira Ticket when clicked on.

The problem is, the uploaded Jira Link does not show in the metadata of the Confluence page itself and neither does it in the linked-to Jira Ticket (Normally, or on every other working Link, the Ticket 'knows' all the pages it is referenced from). If I create the exact same Jira Link again, or I change the key from '12' to '1', save, change back to '12' it works like expected. So why doesn't it in the first place right after the upload?

I have already looked into the source-code of my Confluence page where I put a manually created, working Jira Link. That is what I copied to pass through using asciidoc between the ++++ as shown above.

I also tried passing the link as plain asciidoc link, by putting it as is into the doc, and I tried putting it as html link within ++++.
Both show up as normal hyperlink in the final document, but they do not get converted to a Jira Link. If I copy paste this link on the spot, it automatically transforms into a Jira Macro (a working one).

I also looked into the documentation of doctoolchain. In [publishToConfluence Task documentation](http://doctoolchain.org/docToolchain/v2.0.x/015_tasks/03_task_publishToConfluence.html) I found this:

> **jiraServerId** Stores the Jira server ID that your Confluence instance is connected to. If a value is set, all anchors pointing to a Jira ticket will be replaced by the Confluence Jira macro. To function properly, `jiraRoot` must be configured (see [exportJiraIssues](http://doctoolchain.org/docToolchain/v2.0.x/015_tasks/03_task_exportJiraIssues.html)). Here’s an example:
>
> All files to attach will need to be linked inside the document: link:attachment/myfolder/myfile.json[My API definition]

I do not quite understand this small section. When I try setting 'jiraServerId' in my config (within confluence.with), I get an error when trying to run the publishToConfluence task. The keyword 'jiraServerId' is also not included in the example config in this documentation. And does the last line even belong to the description of **jiraServerId**?
However, the corresponding code does nothing else than replacing a normal link with the source code for a Jira Ticket in Confluence:

```C++
def rewriteJiraLinks = { body ->
    // find links to jira tickets and replace them with jira macros
    body.select('a[href]').each { a ->
        def href = a.attr('href')
        if (href.startsWith(config.jira.api + "/browse/")) {
                def ticketId = a.text()
                a.before("""<ac:structured-macro ac:name=\"jira\" ac:schema-version=\"1\">
                     <ac:parameter ac:name=\"key\">${ticketId}</ac:parameter>
                     <ac:parameter ac:name=\"serverId\">${config.confluence.jiraServerId}</ac:parameter>
                     </ac:structured-macro>""")
                a.remove()
        }
    }
}
```

Which is exactly what I passed manually.

So, can anybody explain to me what is happening here or if I am missing something?

**Comment by DaviSchu** [comment]
Today I wanted to look into this issue again by simply pushing the same page again and waiting for some time, because I thought maybe confluence just needs a brief moment to properly link the tickets when they are introduced this way. I did not change anything in the config, or the corresponding files. Everything is the same as I left it yesterday.
Now, when I try to execute the task `publishToConfluence`, I get the error message
```
* Where:
Script '/Users/david/docToolchain-1.4.0/scripts/publishToConfluence.gradle' line: 27

* What went wrong:
Execution failed for task ':publishToConfluence'.
> No signature of method: groovy.util.ConfigObject.plus() is applicable for argument types: (String) values: [1. Something]
  Possible solutions: plus(java.util.Collection), plus(java.util.Map), values(), put(java.lang.Object, java.lang.Object), split(groovy.lang.Closure), minus(java.util.Map)
```

Where '1. Something' is the only chapter I currently include in my adoc file. It is just an empty page looking like this
```asciidoc
== Something

test test
```

The exact same publish worked several times this week, and suddenly, from closing and opening the laptop, it is broken.
`generateHTML` works as expected, while `generatePDF` gives the same error (which has also worked before).

I get the same error message when using doctoolchain 2.1.0 and 1.4.0, I also used Java 1.8, 11 and 19. Only for 19 I get a different error, but that one just states wrong major version.

Did anything out of my control change while making the leap into October 1st?

My current versions with doctoolchain 1.4.0 are:
```
------------------------------------------------------------
Gradle 6.7.1
------------------------------------------------------------

Build time:   2020-11-16 17:09:24 UTC
Revision:     2972ff02f3210d2ceed2f1ea880f026acfbab5c0

Kotlin:       1.3.72
Groovy:       2.5.12
Ant:          Apache Ant(TM) version 1.10.8 compiled on May 10 2020
JVM:          11.0.2 (Oracle Corporation 11.0.2+9)
OS:           Mac OS X 10.16 x86_64

```

and for 2.1.0:
```
------------------------------------------------------------
Gradle 6.9.2
------------------------------------------------------------

Build time:   2021-12-21 20:18:38 UTC
Revision:     5d94aa68c0fdbe443838bb977080e3b9f273e889

Kotlin:       1.4.20
Groovy:       2.5.12
Ant:          Apache Ant(TM) version 1.10.9 compiled on September 27 2020
JVM:          11.0.2 (Oracle Corporation 11.0.2+9)
OS:           Mac OS X 10.16 x86_64

```


EDIT: With other projects, generatePDF still works, but I do not know why my project suddenly broke

EDIT 2: I setup a new project, where I copied the config and all the other data and the publishToConfluence Task works. I still do not know why my original project became unusable with DocToolchain

EDIT 3: So I pushed the page again on Friday. Today (Monday) the link is still not fully functional

**Reply by rdmueller** [reply]
Sorry that you didn't get any help yet.

regarding the broken toolchain:

- you might want to execute `./dtcw getJava` on your 2.x version. If you have the updated dtc-wrapper in place, this should go and fetch a jdk11 which is only used for docToolchain. This might clarify the Java version
- have you tried to remove the `./build` folder? sometimes files don't get removed there and cause problems.

**Comment by rdmueller** [comment]
the problem you describe sounds to me like a confluence problem. So I don't know if we are able to solve it through docToolchain.
In the past, I only used links to jira and not the confluence-jira macro.
I have just set up a new jira instance and might find some time to play around with it.

**Reply by rdmueller** [reply]
Do you use the cloud or the on premise vresions?
